README.md
[![Build Status](https://travis-ci.com/Huulivoide/format_bytes.svg?branch=master)](https://travis-ci.com/Huulivoide/format_bytes)

Format byte amounts in human readable form.

## Usage

A simple usage example:

```dart
import 'package:format_bytes/format_bytes.dart';

main() {
  print(format(1024)); // 1 KiB
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/Huulivoide/format_bytes/issues


CHANGELOG.md
## 2.1.0
- Actually push to pub.dev
- Update dependencies and enable compatability with dart 3

## 2.0
- Null safety

## 1.0.1
- Some minor improvements to make Dart Pub happier.
  No code changes to the actual library have been made.

## 1.0.0

- Initial version, created by Huulivoide


test/units_test.dart
import 'dart:math';
import 'package:test/test.dart';
import 'package:format_bytes/src/units.dart';

void main() {
  group('Units', () {
    group('find(String byName)', () {
      test('should find megabyte', () {
        var unit = Unit.find('megabyte');
        expect(unit, Unit.megabyte);
      });

      test('should find pebibyte', () {
        var unit = Unit.find('pebibyte');
        expect(unit, Unit.pebibyte);
      });

      test('should not find auto', () {
        expect(() => Unit.find('auto'), throwsArgumentError);
      });
    });

    group('getBest(String for)', () {
      group('Error Conditions', () {
        test('should throw if amount is nonfinite ∞', () {
          expect(() => Unit.bestFor(double.infinity), throwsArgumentError);
        });

        test('should throw if amount is nonfinite -∞', () {
          expect(
              () => Unit.bestFor(double.negativeInfinity), throwsArgumentError);
        });

        test('should throw if amount is NaN', () {
          expect(() => Unit.bestFor(double.nan), throwsArgumentError);
        });
      });

      group('Decimal units', () {
        test('shoud get byte for 0', () {
          var unit = Unit.bestFor(0, UnitType.decimal);
          expect(unit, Unit.byte);
        });

        test('shoud get byte for -0', () {
          var unit = Unit.bestFor(-0, UnitType.decimal);
          expect(unit, Unit.byte);
        });

        const double base = 1000;

        [
          Unit.byte,
          Unit.kilobyte,
          Unit.megabyte,
          Unit.gigabyte,
          Unit.terabyte,
          Unit.petabyte,
          Unit.exabyte,
          Unit.zettabyte,
          Unit.yottabyte
        ].asMap().forEach((i, expectedUnit) {
          test('should return ${expectedUnit.name} for ${base.toInt()}^$i', () {
            var unit = Unit.bestFor(pow(base, i) as double, UnitType.decimal);
            expect(unit.name, expectedUnit.name);
          });
        });
      });

      group('Decimal units', () {
        test('shoud get byte for 0', () {
          var unit = Unit.bestFor(0, UnitType.binary);
          expect(unit, Unit.byte);
        });

        test('shoud get byte for -0', () {
          var unit = Unit.bestFor(-0, UnitType.binary);
          expect(unit, Unit.byte);
        });

        const double base = 1024;

        [
          Unit.byte,
          Unit.kibibyte,
          Unit.mebibyte,
          Unit.gibibyte,
          Unit.tebibyte,
          Unit.pebibyte,
          Unit.exbibyte,
          Unit.zebibyte,
          Unit.yobibyte
        ].asMap().forEach((i, expectedUnit) {
          test('should return ${expectedUnit.name} for ${base.toInt()}^$i', () {
            var unit = Unit.bestFor(pow(base, i) as double, UnitType.binary);
            expect(unit.name, expectedUnit.name);
          });
        });
      });
    });
  });
}


test/localizations_test.dart
import 'package:test/test.dart';
import 'package:format_bytes/src/units.dart';
import 'package:format_bytes/src/localizations.dart';

void main() {
  group('Get Localization', () {
    test('should get real existing one (ru)', () {
      var localization = getLocalization('ru');

      expect(localization, TypeMatcher<Localization>());
      expect(localization.units, containsPair(Unit.mebibyte, 'МиБ'));
    });

    test('should fallback to general version ru-ru → ru', () {
      expect(getLocalization('ru-ru'), getLocalization('ru'));
    });

    test('should fallback to english if requested locale is not found', () {
      expect(getLocalization('non-locale'), getLocalization('en'));
    });
  });
}


test/format_bytes_base_test.dart
import 'dart:math';
import 'package:test/test.dart';
import 'package:format_bytes/src/format_bytes_base.dart';
import 'package:format_bytes/src/units.dart';

void main() {
  group('Format', () {
    group('Handle Error Conditions', () {
      test('should not allow negative value for minDecimals', () {
        expect(
            () => format(1024, minDecimals: -1),
            allOf([
              throwsArgumentError,
              throwsA(
                  wrapMatcher((ArgumentError it) => (it.name == 'minDecimals')))
            ]));
      });

      test('should not allow negative value for maxDecimals', () {
        expect(
            () => format(1024, maxDecimals: -1),
            allOf([
              throwsArgumentError,
              throwsA(
                  wrapMatcher((ArgumentError it) => (it.name == 'maxDecimals')))
            ]));
      });

      test('should not allow maxDecimals value that is less than minDecimals',
          () {
        expect(
            () => format(1024, maxDecimals: 1, minDecimals: 2),
            allOf([
              throwsArgumentError,
              throwsA(
                  wrapMatcher((ArgumentError it) => (it.name == 'maxDecimals')))
            ]));
      });

      test('should not allow negative value (other than -1) for maxPrecision',
          () {
        expect(
            () => format(1024, maxPrecision: -2),
            allOf([
              throwsArgumentError,
              throwsA(wrapMatcher(
                  (ArgumentError it) => (it.name == 'maxPrecision')))
            ]));
      });

      test('should not allow 0 as a value for maxPrecision', () {
        expect(
            () => format(1024, maxPrecision: 0),
            allOf([
              throwsArgumentError,
              throwsA(wrapMatcher(
                  (ArgumentError it) => (it.name == 'maxPrecision')))
            ]));
      });
    });

    group('Handle Good Paramaters', () {
      group('Amount', () {
        test('should format 0 bytes', () {
          expect(format(0), '0 B');
        });

        test('should format -0 bytes', () {
          expect(format(-0.0), '-0 B');
        });

        test('should format 1023 bytes', () {
          expect(format(1023), '1,023 B');
        });

        test('should format 1KiB', () {
          expect(format(1024), '1 KiB');
        });

        test('should format 1MiB', () {
          expect(format(pow(1024, 2)), '1 MiB');
        });

        test('should format 1GiB', () {
          expect(format(pow(1024, 3)), '1 GiB');
        });

        test('should format 1TiB', () {
          expect(format(pow(1024, 4)), '1 TiB');
        });

        test('should format 1PiB', () {
          expect(format(pow(1024, 5)), '1 PiB');
        });

        test('should format 1EiB', () {
          expect(format(pow(1024, 6)), '1 EiB');
        });

        test('should format 1ZiB', () {
          expect(format(pow(1024.0, 7)), '1 ZiB');
        });

        test('should format 1YiB', () {
          expect(format(pow(1024.0, 8)), '1 YiB');
        });
      });

      group('Locale', () {
        test('should use requested locale', () {
          expect(format(1024, locale: 'ru'), '1 КиБ');
        });
      });

      group('FormatString', () {
        test('should replace %amount', () {
          expect(format(1024, formatString: '%amount'), '1');
        });

        test('should replace %sep', () {
          expect(format(1024, formatString: '%sep'), ' ');
        });

        test('should replace %unit', () {
          expect(format(1024, formatString: '%unit'), 'KiB');
        });

        test('should not touch other stuff', () {
          expect(format(1024, formatString: '%amount first %sep second %unit'),
              '1 first   second KiB');
        });
      });

      group('Separator', () {
        test('should use provided custom separator', () {
          expect(format(1024, separator: '😁'), '1😁KiB');
        });
      });

      group('CustomAmountTransformer', () {
        test('should use provided transformer', () {
          expect(format(1024, customAmountTransformer: (b) => 'custom'),
              'custom KiB');
        });
      });

      group('MinDecimals', () {
        test('should include unneeded decimals if > 0', () {
          expect(format(1024, minDecimals: 2), '1.00 KiB');
        });
      });

      group('MaxDecimals', () {
        test('should cut off all decimals if 0', () {
          expect(format(1500, maxDecimals: 0), '1 KiB');
        });
      });

      group('MaxPrecission', () {
        test('should limit to intgers only', () {
          expect(format(1024 * 123, maxPrecision: 1), '100 KiB');
        });

        test('should limit decimals', () {
          expect(format(1024 * 1.25, maxPrecision: 2), '1.3 KiB');
        });
      });

      group('Unit', () {
        test('should use provided unit', () {
          expect(format((1024 * 1024) / 2, unit: Unit.mebibyte), '0.5 MiB');
        });

        test('should use provided custom unit', () {
          expect(
              format(500, unit: Unit('custom', 10, UnitType.decimal)), '50 ');
        });
      });

      group('UnitType', () {
        test('should use binary units', () {
          expect(format(1024, unitType: UnitType.binary), '1 KiB');
        });

        test('should use decimal units', () {
          expect(format(1000, unitType: UnitType.decimal), '1 KB');
        });
      });
    });
  });
}


lib/src/localizations.dart
import 'units.dart';

class Localization {
  final String format;
  final String separator;
  final Map<Unit, String> units;

  const Localization(this.format, this.separator, this.units);
}

const _localizedUnits = {
  'en': Localization('%amount%sep%unit', ' ', {
    Unit.byte: 'B',
    Unit.kilobyte: 'KB',
    Unit.megabyte: 'MB',
    Unit.gigabyte: 'GB',
    Unit.terabyte: 'TB',
    Unit.petabyte: 'PB',
    Unit.exabyte: 'EB',
    Unit.zettabyte: 'ZB',
    Unit.yottabyte: 'YB',
    Unit.kibibyte: 'KiB',
    Unit.mebibyte: 'MiB',
    Unit.gibibyte: 'GiB',
    Unit.tebibyte: 'TiB',
    Unit.pebibyte: 'PiB',
    Unit.exbibyte: 'EiB',
    Unit.zebibyte: 'ZiB',
    Unit.yobibyte: 'YiB'
  }),
  'fi': Localization('%amount%sep%unit', ' ', {
    Unit.byte: 't',
    Unit.kilobyte: 'Kt',
    Unit.megabyte: 'Mt',
    Unit.gigabyte: 'Gt',
    Unit.terabyte: 'Tt',
    Unit.petabyte: 'Pt',
    Unit.exabyte: 'Et',
    Unit.zettabyte: 'Zt',
    Unit.yottabyte: 'Yt',
    Unit.kibibyte: 'KiB',
    Unit.mebibyte: 'MiB',
    Unit.gibibyte: 'GiB',
    Unit.tebibyte: 'TiB',
    Unit.pebibyte: 'PiB',
    Unit.exbibyte: 'EiB',
    Unit.zebibyte: 'ZiB',
    Unit.yobibyte: 'YiB'
  }),
  'ru': Localization('%amount%sep%unit', ' ', {
    Unit.byte: 'б',
    Unit.kilobyte: 'Кб',
    Unit.megabyte: 'Мб',
    Unit.gigabyte: 'Гб',
    Unit.terabyte: 'Тб',
    Unit.petabyte: 'Пб',
    Unit.exabyte: 'Эб',
    Unit.zettabyte: 'Зб',
    Unit.yottabyte: 'Йб',
    Unit.kibibyte: 'КиБ',
    Unit.mebibyte: 'МиБ',
    Unit.gibibyte: 'ГиБ',
    Unit.tebibyte: 'ТиБ',
    Unit.pebibyte: 'ПиБ',
    Unit.exbibyte: 'ЭиБ',
    Unit.zebibyte: 'ЗиБ',
    Unit.yobibyte: 'ЙиБ'
  }),
};

Localization getLocalization(String locale) {
  if (_localizedUnits.containsKey(locale)) return _localizedUnits[locale]!;

  var primary = locale.split('-')[0];
  if (_localizedUnits.containsKey(primary)) return _localizedUnits[primary]!;

  return _localizedUnits['en']!;
}


lib/src/format_bytes_base.dart
import 'package:intl/intl.dart' show NumberFormat;
import 'units.dart';
import 'localizations.dart';

const defaultFormatString = 'locale default';
const defaultSeparator = defaultFormatString;

/// A function that transforms [double] to [String].
///
/// The build in [double.toString] can result in ouputs like 1.123e+24 which
/// might be unwanted.
typedef AmountToString = String Function(double amount);

/// Creates a [AmountToString] transformer with specified precision.
///
/// Creates a [NumberFormat] based transformer function.
///
/// ```
/// def fmt = createPrecisionFormatter(0, 0, 3, 'en-us');
/// print(fmt(123456.789)); // → '123000'
/// ```
///
/// [minDecimals] ans [maxDecimals] specifies the minimum and maximum amount of
/// decimal places included in the output. [maxPrecision] defines the maximum
/// significant digits included in the output. [maxPrecision] can be disabled
/// with -1. Finally [locale] can be used to control the number styling
/// and decimal separator.
///
/// Will throw [ArgumentError] if the following conditions are not met.
/// * 0 <= [minDecimals] <= [maxDecimals]
/// * [maxPrecision] == -1 or 1 <= [maxPrecision]
///
AmountToString createPrecisionFormatter(
    int minDecimals, int maxDecimals, int maxPrecision, String locale) {
  if (minDecimals < 0) {
    throw ArgumentError.value(
        minDecimals, 'minDecimals', 'Expected positive value.');
  }

  if (maxDecimals < 0 || maxDecimals < minDecimals) {
    throw ArgumentError.value(maxDecimals, 'maxDecimals',
        'Expected positive value larger or equal to minDecimals.');
  }

  if (maxPrecision < -1 || maxPrecision == 0) {
    throw ArgumentError.value(maxPrecision, 'maxPrecision',
        'Expected positive non zero value or -1 (no limit).');
  }

  var fmt = NumberFormat.decimalPattern(locale)
    ..minimumFractionDigits = minDecimals
    ..maximumFractionDigits = maxDecimals
    ..significantDigits = maxPrecision
    ..significantDigitsInUse = (maxPrecision > 0);

  return (double amount) {
    return fmt.format(amount);
  };
}

Unit _resolveRealUnit(double amount, Unit possibleUnit, UnitType type) {
  if (possibleUnit != Unit.auto) return possibleUnit;

  return Unit.bestFor(amount, type);
}

String _formatString(double amount, AmountToString amountToString,
    String format, String separator, String unit) {
  return format
      .replaceFirst('%amount', amountToString(amount))
      .replaceFirst('%sep', separator)
      .replaceFirst('%unit', unit);
}

String _formatSafeParams(
    num amount,
    String locale,
    String formatString,
    String separator,
    AmountToString? customAmountTransformer,
    int minDecimals,
    int maxDecimals,
    int maxPrecision,
    Unit unit,
    UnitType unitType) {
  var doubleAmount = amount.toDouble();
  var amountToString = customAmountTransformer ??
      createPrecisionFormatter(minDecimals, maxDecimals, maxPrecision, locale);

  var realUnit = _resolveRealUnit(doubleAmount, unit, unitType);
  var localization = getLocalization(locale);

  return _formatString(
      doubleAmount / realUnit.multiplier,
      amountToString,
      (formatString == defaultFormatString)
          ? localization.format
          : formatString,
      (separator == defaultSeparator) ? localization.separator : separator,
      localization.units[realUnit] ?? '');
}

/// Formats a byte amount as human readable string.
///
/// The only required parameter is [amount], which is a number describing a
/// amount of bytes. That amount is then formatted with appropriate binary
/// [Unit] as determined by [Unit.bestFor]. If one prefers one can get a decimal
/// based result by providing [UnitType.decimal] as the [unitType] parameter.
///
/// ```
/// String output = format(3 * pow(1024, 3)) // 3 GiB
/// ```
///
/// Simplest way to localize the output is to provide the function with the
/// [locale] parameter. Currently supported locales are en, fi and ru.
///
/// ```
/// String output = format(3 * pow(1024, 3), locale: 'ru') // 3 ГиБ
/// ```
///
/// Output precision can be controlled with the parameters [minDecimals],
/// [maxDecimals] and [maxPrecision] which will be passed on to
/// [createPrecisionFormatter].
///
/// ```
/// String output = format(1024, minDecimals: 1) // 1,0 KiB
/// ```
///
/// To use a specific unit instead of automatically determined one, you can
/// provide the wanted unit as the [unit] parameter.
///
/// ```
/// String output = format(1024, unit: Unit.byte) // 1024 B
/// ```
///
/// ## Custom formatting
///
/// It is also possible to define the formatting manually with [formatString],
/// [separator] and [customAmountTransformer]. [formatString] can contain the
/// following placeholders _%amount_, _%sep_ and _%unit_. [separator] overrides
/// the default separator character used between the amount and the unit.
///
/// ```
/// String output = format(1024,
///   formatString: '%unit%sep%amount',
///   separator: '__') // KiB__1
/// ```
///
/// Providing a customAmountTransformer will override the default
/// [createPrecisionFormatter] number to string transformer function.
///
/// ```
/// String output = format(12.345e67,
///   customAmountTransformer: (amount) { amount.toString() },
///   unit: Unit.byte
/// ) // 1.2345e+68 B
/// ```
///
/// Custom unit can be used by providing a format string with the %unit
/// placeholder hard coded in. In addition a custom [Unit] can be provided
/// as the [unit] parameter.
///
/// ```
/// String output = format(1024,
///   formatString: '%amount half-KiB',
///   unit: Unit('half-kibibyte', 512, UnitType.binary)) // 2 half-KiB
/// ```
///
/// Will throw [ArgumentError] if any of the provided parameters is *null*,
/// customAmountTransformer is a an exception and it can be and does default
/// to null.
///
String format(num amount,
    {String locale = 'en',
    String formatString = defaultFormatString,
    String separator = defaultSeparator,
    AmountToString? customAmountTransformer,
    int minDecimals = 0,
    int maxDecimals = 2,
    int maxPrecision = -1,
    Unit unit = Unit.auto,
    UnitType unitType = UnitType.binary}) {
  return _formatSafeParams(
      amount,
      locale,
      formatString,
      separator,
      customAmountTransformer,
      minDecimals,
      maxDecimals,
      maxPrecision,
      unit,
      unitType);
}


lib/src/units.dart
/// Binary ot Decimal based unit
///
/// Used to determine the type of [Unit] and to select appropriate unit type.
enum UnitType { binary, decimal }

/// Describes a storage amount unit
///
/// While a public constructor is provided you should really consider this as a
/// enum class instead.
///
/// ## Provided values:
///
/// * [Unit.auto]: Default, special, unit, which instructs the formatting code to
/// chose automatically choose the best unit based on the provided inputs.
/// * [Unit.byte]: While it is marked as a decimal based unit it does also refer
/// to binary based unit.
///
/// ### Decimal (1000-based multiplier) units
///  * [Unit.kilobyte]: 1 KB
///  * [Unit.megabyte]: 1 MB
///  * [Unit.gigabyte]: 1 GB
///  * [Unit.petabyte]: 1 TB
///  * [Unit.exabyte]: 1 EB
///  * [Unit.zettabyte]: 1 ZB
///  * [Unit.yottabyte]: 1 YB
///  * [Unit.kilobyte]: 1 KB
///
/// ### Binary (1024-based multiplier) units
///  * [Unit.kibibyte]: 1 KiB
///  * [Unit.mebibyte]: 1 MiB
///  * [Unit.gibibyte]: 1 GiB
///  * [Unit.tebibyte]: 1 TiB
///  * [Unit.pebibyte]: 1 PiB
///  * [Unit.exbibyte]: 1 EiB
///  * [Unit.zebibyte]: 1 ZiB
///  * [Unit.yobibyte]: 1 YiB
class Unit {
  final String name;
  final double multiplier;
  final UnitType type;

  /// Primary constructor
  ///
  /// Defines the units behaviour and identity. [name] is the name of the unit.
  /// [type] is one of the [UnitType] types, for custom units this has no
  /// functional effect. The primary parameter is [multiplier] and it is the one
  /// that is actually used by the formatting code to generate the final amount
  /// from the input. It is the number, ny which one must divide a input number,
  /// to get a appropriate amount for this particular unit. For example 1024,
  /// in the case of [Unit.kibibyte].
  const Unit(this.name, this.multiplier, this.type);

  static const auto = Unit('auto', 0, UnitType.decimal);
  static const byte = Unit('byte', 1, UnitType.decimal);

  static const kilobyte = Unit('kilobyte', 1000, UnitType.decimal);
  static const megabyte = Unit('megabyte', 1000000, UnitType.decimal);
  static const gigabyte = Unit('gigabyte', 1000000000, UnitType.decimal);
  static const terabyte = Unit('terabyte', 1000000000000, UnitType.decimal);
  static const petabyte = Unit('petabyte', 1000000000000000, UnitType.decimal);
  static const exabyte = Unit('exabyte', 1000000000000000000, UnitType.decimal);
  static const zettabyte =
      Unit('zettabyte', 1000000000000000000000, UnitType.decimal);
  static const yottabyte = Unit(
      'yottabyte',
      // Suppress analyzer error (with .0) about possible integer overflow.
      // There is no overflow in this particular case.
      1000000000000000000000000.0,
      UnitType.decimal);

  static const kibibyte = Unit('kibibyte', 1024, UnitType.binary);
  static const mebibyte = Unit('mebibyte', 1048576, UnitType.binary);
  static const gibibyte = Unit('gibibyte', 1073741824, UnitType.binary);
  static const tebibyte = Unit('tebibyte', 1099511627776, UnitType.binary);
  static const pebibyte = Unit('pebibyte', 1125899906842624, UnitType.binary);
  static const exbibyte =
      Unit('exbibyte', 1152921504606846976, UnitType.binary);
  static const zebibyte =
      Unit('zebibyte', 1180591620717411303424, UnitType.binary);
  static const yobibyte =
      Unit('yobibyte', 1208925819614629174706176, UnitType.binary);

  static const _binaryUnits = [
    Unit.yobibyte,
    Unit.zebibyte,
    Unit.exbibyte,
    Unit.pebibyte,
    Unit.tebibyte,
    Unit.gibibyte,
    Unit.mebibyte,
    Unit.kibibyte,
    Unit.byte
  ];

  static const _decimalUnits = [
    Unit.yottabyte,
    Unit.zettabyte,
    Unit.exabyte,
    Unit.petabyte,
    Unit.terabyte,
    Unit.gigabyte,
    Unit.megabyte,
    Unit.kilobyte,
    Unit.byte
  ];

  static const _byName = {
    'byte': Unit.byte,
    'kilobyte': Unit.kilobyte,
    'megabyte': Unit.megabyte,
    'gigabyte': Unit.gigabyte,
    'terabyte': Unit.terabyte,
    'petabyte': Unit.petabyte,
    'exabyte': Unit.exabyte,
    'zettabyte': Unit.zettabyte,
    'yottabyte': Unit.yottabyte,
    'kibibyte': Unit.kibibyte,
    'mebibyte': Unit.mebibyte,
    'gibibyte': Unit.gibibyte,
    'tebibyte': Unit.tebibyte,
    'pebibyte': Unit.pebibyte,
    'exbibyte': Unit.exbibyte,
    'zebibyte': Unit.zebibyte,
    'yobibyte': Unit.yobibyte
  };

  /// Find a predefined unit by its name.
  ///
  /// ```
  ///  double mbMultiplier = Unit.find('megabyte').multiplier;
  /// ```
  ///
  /// While this function is provided, you should, if possible, refer directly
  /// to the wanted unit.
  ///
  /// [name] is the name field of the requested unit.
  /// Throws [ArgumentError] if the requested unit is not found.
  static Unit find(String name) {
    var unit = _byName[name];

    if (unit == null) {
      throw ArgumentError.value(name, 'name',
          'Expected name to be one of ${_byName.keys.join(', ')}.');
    }

    return unit;
  }

  static List<Unit> _getUnitListForType(UnitType type) {
    switch (type) {
      case UnitType.binary:
        return _binaryUnits;
      case UnitType.decimal:
        return _decimalUnits;
    }
  }

  /// Determines the best unit for given amount and type.
  ///
  /// ```
  /// Unit shouldBeKibibyte = Unit.bestFor(2000, UnitType.binary);
  /// ```
  ///
  /// [amount] is the only required parameter. If [type] is not provided,
  /// will default to returning a binary based unit.
  ///
  /// Will throw [ArgumentError] if amount is a non finite value. That is
  /// [double.infinity], [double.negativeInfinity] or [double.nan].
  static Unit bestFor(double amount, [UnitType type = UnitType.binary]) {
    if (!amount.isFinite) {
      throw ArgumentError.value(amount, 'amount', 'Expected a finite value.');
    }

    var positive = amount.abs();

    if (positive == 0) {
      return Unit.byte;
    }

    return _getUnitListForType(type)
        .firstWhere((unit) => (unit.multiplier <= positive));
  }
}
